### PR TITLE
(BIDS-1671) (BIDS-1677) withrawal tab sorting + adding to dash

### PR DESCRIPTION
**Pull request by MauserBitfly**
With this PR:
- add new withdrawal tab to dashboard
- fix sorting of withdrawal tab on Validator page
- deactivate sorting on attestation tab on Validator page

### CODE REVIEW

File: handlers/dashboard.go [hunk 1]
```diff
@@ -344,49 +349,89 @@ func DashboardDataProposals(w http.ResponseWriter, r *http.Request) {
...
+		RecordsTotal:    withdrawalCount,
+		RecordsFiltered: withdrawalCount,
+		Data:            tableData,
+	}
+
+	err = json.NewEncoder(w).Encode(data)
 	if err != nil {
-		logger.WithError(err).WithField("route", r.URL.String()).Error("error enconding json response")
-		http.Error(w, "Internal server error", http.StatusServiceUnavailable)
+		logger.Errorf("error enconding json response for %v route: %v", r.URL.String(), err)
```
Comment (recy21) [inline handlers/dashboard.go:433]: It's better practice to use fields, than having everything in one string. But you after the merge of BIDS-1297, I would suggest to always use utils.LogError / utils.LogFatal, which will do all that for you.

File: handlers/dashboard.go [hunk 2]
```diff
@@ -344,49 +349,89 @@ func DashboardDataProposals(w http.ResponseWriter, r *http.Request) {
...
 
-	proposalsResult := make([][]uint64, len(proposals))
-	for i, b := range proposals {
-		proposalsResult[i] = []uint64{
-			uint64(utils.SlotToTime(b.Slot).Unix()),
-			b.Status,
-		}
+	withdrawals, err := db.GetDashboardWithdrawals(validators, length, start, orderBy, orderDir)
+	if err != nil {
+		logger.Errorf("error retrieving validator withdrawals: %v", err)
```
Comment (recy21) [inline handlers/dashboard.go:406]: I would suggest to always use utils.LogError / utils.LogFatal, which will do all that for you.

File: handlers/dashboard.go [hunk 3]
```diff
@@ -344,49 +349,89 @@ func DashboardDataProposals(w http.ResponseWriter, r *http.Request) {
...
-	}{}
+	draw, err := strconv.ParseUint(q.Get("draw"), 10, 64)
+	if err != nil {
+		logger.Errorf("error converting datatables data parameter from string to int: %v", err)
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		return
+	}
+	start, err := strconv.ParseUint(q.Get("start"), 10, 64)
+	if err != nil {
+		logger.Errorf("error converting datatables start parameter from string to int: %v", err)
```
Comment (recy21) [inline handlers/dashboard.go:372]: I would suggest to always use utils.LogError / utils.LogFatal, which will do all that for you.

File: handlers/dashboard.go [hunk 4]
```diff
@@ -344,49 +349,89 @@ func DashboardDataProposals(w http.ResponseWriter, r *http.Request) {
...
 	}
-	filter := pq.Array(filterArr)
 
-	proposals := []struct {
-		Slot   uint64
-		Status uint64
-	}{}
+	draw, err := strconv.ParseUint(q.Get("draw"), 10, 64)
+	if err != nil {
+		logger.Errorf("error converting datatables data parameter from string to int: %v", err)
```
Comment (recy21) [inline handlers/dashboard.go:366]: I would suggest to always use utils.LogError / utils.LogFatal, which will do all that for you.

File: handlers/dashboard.go [hunk 5]
```diff
@@ -344,49 +349,89 @@ func DashboardDataProposals(w http.ResponseWriter, r *http.Request) {
...
+		orderDir = "desc"
+	}
+
+	length := uint64(10)
+
+	withdrawalCount, err := db.GetDashboardWithdrawalsCount(validators)
 	if err != nil {
-		logger.WithError(err).WithField("route", r.URL.String()).Error("error retrieving block-proposals")
-		http.Error(w, "Internal server error", http.StatusServiceUnavailable)
+		logger.Errorf("error retrieving dashboard validator withdrawals count: %v", err)
```
Comment (recy21) [inline handlers/dashboard.go:399]: I would suggest to always use utils.LogError / utils.LogFatal, which will do all that for you.

**Review by recy21 — CHANGES_REQUESTED** [review]

File: handlers/dashboard.go (hunk 4 above)
Comment (MauserBitfly) [reply handlers/dashboard.go:366]: replaced with utils.LogError

File: handlers/dashboard.go (hunk 3 above)
Comment (MauserBitfly) [reply handlers/dashboard.go:372]: replaced with utils.LogError

File: handlers/dashboard.go (hunk 5 above)
Comment (MauserBitfly) [reply handlers/dashboard.go:399]: replaced with utils.LogError

**Review by recy21 — APPROVED** [review]

**Review by D13ce — CHANGES_REQUESTED** [review]
According to BIDS-1677, the sorting on the validator page should only be enabled for `slot` and `amount`. Currently the validator page lets you sort for all columns. Please change.

This change has to be done for the dashboard as well.

In some of my comments I suggest to change something to `epoch`, this the would be wrong of course since `epoch`s can not be sorted anymore.

File: templates/dashboard/tables.html [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+      if (isDashboardActive()) {
+        createWithdrawalTable()
+      }
+
+      $('a[data-toggle="tab"]').on("shown.bs.tab", function (e) {
+        if (e.target.id === "withdrawal-tab" && !table) {
+          createWithdrawalTable()
+        }
+      })
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply templates/dashboard/tables.html:43]: When you select the withdrawal tab for the first time the data is loaded

File: handlers/dashboard.go [hunk 7]
```diff
@@ -344,49 +349,89 @@ func DashboardDataProposals(w http.ResponseWriter, r *http.Request) {
...
 
-	err = db.ReaderDb.Select(&proposals, `
-		SELECT slot, status
-		FROM blocks
-		WHERE proposer = ANY($1)
-		ORDER BY slot`, filter)
+	orderColumn := q.Get("order[0][column]")
+	orderByMap := map[string]string{
+		"0": "validatorindex",
+		"1": "block_slot",
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply handlers/dashboard.go:380]: No there is no DB field 'epoch' ... the 'epoch' is calculated for the ui based on the slot, so for db sorting its the same if you sort for epoch or slot (or time)

File: templates/dashboard/tables.html [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+      var validatorsStr = (validatorsStr = localStorage.getItem("dashboard_validators"))
+
+      if (!validatorsStr) {
+        var usp = new URLSearchParams(window.location.search)
+        validatorsStr = usp.get("validators")
+      }
+      return validatorsStr ? JSON.parse(validatorsStr).join(",") : undefined
+    }
+
+    function isDashboardActive() {
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply templates/dashboard/tables.html]: renamed

File: handlers/validator.go [hunk 9]
```diff
@@ -1151,6 +1151,23 @@ func ValidatorWithdrawals(w http.ResponseWriter, r *http.Request) {
...
+	orderByMap := map[string]string{
+		"0": "block_slot",
+		"1": "block_slot",
+		"2": "withdrawalindex",
+		"3": "address",
+		"4": "amount",
+	}
+	orderBy, exists := orderByMap[orderColumn]
+	if !exists {
+		orderBy = "validatorindex"
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply handlers/validator.go:1164]: replaced with block_slot

File: db/db.go [hunk 10]
```diff
@@ -2424,6 +2424,51 @@ func GetAddressWithdrawalsTotal(address []byte) (uint64, error) {
...
+	SELECT 
+		w.block_slot as slot, 
+		w.withdrawalindex as index, 
+		w.validatorindex, 
+		w.address, 
+		w.amount 
+	FROM blocks_withdrawals w
+	INNER JOIN blocks b ON b.blockroot = w.block_root AND b.status = '1'
+	WHERE validatorindex = ANY($1)
+	ORDER BY  w.`+orderBy+` `+orderDir+` LIMIT $2 OFFSET $3`, validatorFilter, limit, offset)
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply db/db.go]: replaced with  fmt.Sprintf(`

**Review by D13ce — APPROVED** [review]
lgtm

**Pull request merged**